### PR TITLE
Add a warning for potentially needless use of setSecretKey

### PR DESCRIPTION
**Pull request by AdamBJohnsonx**

### CODE REVIEW

**Review by tanmaymanolkar1 — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java [hunk 1]
```diff
@@ -108,11 +112,15 @@
...
      */
-    public void setSecretKey(byte[] rawKey) {
+    public synchronized void setSecretKey(byte[] rawKey) {
         if (rawKey == null || rawKey.length != SECRET_RAW_KEY_LENGTH) {
             throw new IllegalArgumentException("rawKey");
         }
-
+        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            Logger.warn(":setSecretKey", "You're using setSecretKey in a version of android" +
+                    "that supports keyStore functionality.  Consider not doing this, as it only exists" +
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:121]:
```suggestion
                    "that supports keyStore functionality.  Consider not doing this, as it only exists " +
```

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java [hunk 2]
```diff
@@ -108,11 +112,15 @@
...
      * @param rawKey App related key to use in encrypt/decrypt
      */
-    public void setSecretKey(byte[] rawKey) {
+    public synchronized void setSecretKey(byte[] rawKey) {
         if (rawKey == null || rawKey.length != SECRET_RAW_KEY_LENGTH) {
             throw new IllegalArgumentException("rawKey");
         }
-
+        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            Logger.warn(":setSecretKey", "You're using setSecretKey in a version of android" +
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:120]:
```suggestion
            Logger.warn(":setSecretKey", "You're using setSecretKey in a version of android " +
```

**Comment by AdamBJohnsonx** [comment]
This is in relation to https://portal.microsofticm.com/imp/v3/incidents/details/210581721

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java [hunk 3]
```diff
@@ -108,11 +112,15 @@
      *
      * @param rawKey App related key to use in encrypt/decrypt
      */
-    public void setSecretKey(byte[] rawKey) {
+    public synchronized void setSecretKey(byte[] rawKey) {
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:115]:
```suggestion
    public void setSecretKey(byte[] rawKey) {
```

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java [hunk 4]
```diff
@@ -112,7 +116,11 @@ public void setSecretKey(byte[] rawKey) {
         if (rawKey == null || rawKey.length != SECRET_RAW_KEY_LENGTH) {
             throw new IllegalArgumentException("rawKey");
         }
-
+        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            Logger.warn(":setSecretKey", "You're using setSecretKey in a version of android " +
+                    "that supports keyStore functionality.  Consider not doing this, as it only exists " +
+                    "for devices with an SDK lower than " + Build.VERSION_CODES.JELLY_BEAN_MR2);
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:122]: and are we doing this right now?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:122]: We support generating our own key, it happens here:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/9a09e2e4a1fe714773a2b679d054b311d2c56f86/common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java#L666

From a customer investigation we were working on earlier, we found an app that was calling this method at startup: we suspect it's causing issues with SSO loss due to the host-app [potentially] providing mismatched keys

Comment (shahzaibj) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:122]: Cool.....actually how does user defined key currently work for MSAL? It is only for ADAL right?

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:131]: MSAL reads the value from the AuthenticationSettings.INSTANCE:

        // Try to get user defined key (ADAL/MSAL).
        if (AuthenticationSettings.INSTANCE.getSecretKeyData() != null) {
            setBlobVersion(VERSION_USER_DEFINED);
            return loadSecretKey(KeyType.ADAL_USER_DEFINED_KEY);
        }

Adam.

**Review by iambmelt — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java (hunk 4 above)
Comment (shahzaibj) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:131]: Yeah but it has to be set on `AuthenticationSettings.INSTANCE` right? ADAL has a public method to do this...but I don't see one in MSAL

Comment (shahzaibj) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:131]: This is ADAL: https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/cfdb40c5d0fb28e4825df030e909bc493aaaffe9/adal/src/main/java/com/microsoft/aad/adal/AuthenticationSettings.java#L82

Is there something similar in MSAL?

Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:131]: PublicClientApplicationConfiguration.java:167

Comment (shahzaibj) [reply common/src/main/java/com/microsoft/identity/common/adal/internal/AuthenticationSettings.java:131]: Cool, Thanks!

**Pull request merged**